util: Bump GPUSE docker to GCC 10

abmerop opened the pull request:
The docker image uses GCC 8 by default which is no longer supported in gem5. Bump the version to GCC 10.

Change-Id: I0e46e0f82d881fe73eccef75e2f343b41073f6fe

BobbyRBruce commented on util/dockerfiles/gcn-gpu/Dockerfile:
```diff
@@ -31,6 +31,10 @@ RUN apt -y update && apt -y upgrade && \
     libhdf5-serial-dev python3-pydot libpng-dev libelf-dev pkg-config gdb
 
+# gem5 requires gcc 10+
+RUN apt -y install gcc-10 g++-10 cpp-10
```
Could you update this to:

```suggestion
RUN apt -y update && apt -y install gcc-10 g++-10 cpp-10
```

Otherwise, this change seems fine to me.

abmerop replied:
There is an apt update on the line directly above

BobbyRBruce replied:
If those lines are cached, this line can fail. This is a common problem when dealing with Dockerfiles.

abmerop replied:
Make sense.  Should I add that to *all* of the apt installs while I'm updating this file?

BobbyRBruce replied:
I don't think that's necessary right now. Adding `apt update -y` is really just a safety. If you encounter problems without it you can just clear the local caches and start from scratch.

BobbyRBruce commented:
I'm getting the following error trying to build this image... I don't fully understand as it has nothing to do with the compiler version increase as far as I can tell:

```sh
 => [gcn-gpu  8/42] RUN echo 'deb [arch=amd64] https://repo.radeon.com/rocm/apt/4.0.1/ xenial main' | tee /etc/apt/sources.list.d/rocm.list                                                                 0.0s 
 => ERROR [gcn-gpu  9/42] RUN apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev                                                                                                                  1.7s 
------                                                                                                                                                                                                           
 > [gcn-gpu  9/42] RUN apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev:                                                                                                                             
0.084 Get:1 https://repo.radeon.com/rocm/apt/4.0.1 xenial InRelease [1817 B]                                                                                                                                     
0.125 Get:2 https://repo.radeon.com/rocm/apt/4.0.1 xenial/main amd64 Packages [17.6 kB]                                                                                                                          
0.354 Hit:3 http://ports.ubuntu.com/ubuntu-ports focal InRelease                                                                                                                                                 
0.494 Hit:4 http://ports.ubuntu.com/ubuntu-ports focal-updates InRelease                                                                                                                                         
0.641 Hit:5 http://ports.ubuntu.com/ubuntu-ports focal-backports InRelease
0.786 Hit:6 http://ports.ubuntu.com/ubuntu-ports focal-security InRelease
0.831 Fetched 19.4 kB in 1s (24.8 kB/s)
0.831 Reading package lists...
1.188 
1.188 WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
1.188 
1.205 Reading package lists...
1.557 Building dependency tree...
1.632 Reading state information...
1.641 E: Unable to locate package hsakmt-roct
1.641 E: Unable to locate package hsakmt-roct-dev
------
failed to solve: process "/bin/sh -c apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev" did not complete successfully: exit code: 100
```

abmerop committed:
util: Bump GPUSE docker to GCC 10

The docker image uses GCC 8 by default which is no longer supported in
gem5. Bump the version to GCC 10.

Change-Id: I0e46e0f82d881fe73eccef75e2f343b41073f6fe

abmerop commented:
> I'm getting the following error trying to build this image... I don't fully understand as it has nothing to do with the compiler version increase as far as I can tell:
> 
> ```shell
>  => [gcn-gpu  8/42] RUN echo 'deb [arch=amd64] https://repo.radeon.com/rocm/apt/4.0.1/ xenial main' | tee /etc/apt/sources.list.d/rocm.list                                                                 0.0s 
>  => ERROR [gcn-gpu  9/42] RUN apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev                                                                                                                  1.7s 
> ------                                                                                                                                                                                                           
>  > [gcn-gpu  9/42] RUN apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev:                                                                                                                             
> 0.084 Get:1 https://repo.radeon.com/rocm/apt/4.0.1 xenial InRelease [1817 B]                                                                                                                                     
> 0.125 Get:2 https://repo.radeon.com/rocm/apt/4.0.1 xenial/main amd64 Packages [17.6 kB]                                                                                                                          
> 0.354 Hit:3 http://ports.ubuntu.com/ubuntu-ports focal InRelease                                                                                                                                                 
> 0.494 Hit:4 http://ports.ubuntu.com/ubuntu-ports focal-updates InRelease                                                                                                                                         
> 0.641 Hit:5 http://ports.ubuntu.com/ubuntu-ports focal-backports InRelease
> 0.786 Hit:6 http://ports.ubuntu.com/ubuntu-ports focal-security InRelease
> 0.831 Fetched 19.4 kB in 1s (24.8 kB/s)
> 0.831 Reading package lists...
> 1.188 
> 1.188 WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
> 1.188 
> 1.205 Reading package lists...
> 1.557 Building dependency tree...
> 1.632 Reading state information...
> 1.641 E: Unable to locate package hsakmt-roct
> 1.641 E: Unable to locate package hsakmt-roct-dev
> ------
> failed to solve: process "/bin/sh -c apt-get update && apt -y install hsakmt-roct hsakmt-roct-dev" did not complete successfully: exit code: 100
> ```

Does it fail every time? I am not seeing this.  I also do not see ubuntu-ports being used though.

mattsinc reviewed (dismissed):
From my perspective this looks ok, but will let @BobbyRBruce sign off since there is a bug I can't seem to replicate

BobbyRBruce commented:
I got past the past that blocked me before... I think maybe i was building on a Apple-silcon host and that may have caused issue though I'm getting the following error now since running on an AMD machine:


```shell
4.789 [ 37%] Building C object khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/callbacks.c.o
4.849 [ 40%] Building C object khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/test_create_calls.c.o
4.989 [ 40%] Building C object khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/test_clgl.c.o
5.093 [ 43%] Building C object khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/test_image_objects.c.o
5.190 [ 43%] Linking C executable icd_loader_test
5.209 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_platforms.c.o:(.bss+0x110): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.214 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_program_objects.c.o:(.bss+0x0): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.215 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_sampler_objects.c.o:(.bss+0x0): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.219 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_cl_runtime.c.o:(.bss+0x0): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.219 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_clgl.c.o:(.bss+0x248): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.219 /usr/bin/ld: CMakeFiles/icd_loader_test.dir/test_image_objects.c.o:(.bss+0x0): multiple definition of `ret_val'; CMakeFiles/icd_loader_test.dir/test_kernel.c.o:(.bss+0x348): first defined here
5.240 collect2: error: ld returned 1 exit status
5.242 make[2]: *** [khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/build.make:251: khronos/icd/test/loader_test/icd_loader_test] Error 1
5.243 make[1]: *** [CMakeFiles/Makefile2:1427: khronos/icd/test/loader_test/CMakeFiles/icd_loader_test.dir/all] Error 2
5.244 make: *** [Makefile:130: all] Error 2
------
failed to solve: process "/bin/sh -c cmake -DUSE_COMGR_LIBRARY=ON -DCMAKE_PREFIX_PATH=\"/opt/rocm\"     -DCMAKE_BUILD_TYPE=Release .. &&     make -j$(nproc) && make package" did not complete successfully: exit code: 2
```

BobbyRBruce commented:
Appears it's this issue: https://github.com/rocm-arch/rocm-arch/issues/165

BobbyRBruce committed:
util-docker: Patch 'mult def of ret_val' Linker error

Issue highlighted here:
https://github.com/ROCm/ROCm-OpenCL-Runtime/issues/113

This patch applies the patch to ROCm-OpenCL-Runtime in our gcn-gpu
Dockerfile

Change-Id: I71870924fa9be08883db57f6552d5268e8e14649

BobbyRBruce commented:
I pushed a commit to this PR which stops the linking error I've been encountering when trying to build ROCm-OpenCL-Runtime. I'm now building the image again. Hopefully i can get it pushed soon.

BobbyRBruce commented:
With my change i was able to build the image. I pushed it to the docker registry. This _should_ fix the failing tests. We should know by tomorrow.

abmerop commented:
Thanks for looking into this!  It seems likely we'd want to go with #1170 instead. 

I am a little concerned that as this code ages more the latest compilers will fail in more ways and more and more hacky patches will be needed.  My current plan is to have some kind of APU working in full system mode (i.e., CPU+GPU memory are shared).  Right now this is the only type of GPU that can't currently be simulated in GPUFS.  After that, SE mode wouldn't have many use cases and this docker image could potentially be deprecated.

powerjg commented:
> I am a little concerned that as this code ages more the latest compilers will fail in more ways and more and more hacky patches will be needed.

Me too! Thanks for working on this and helping out, Matt!

BobbyRBruce commented:
This change has been incorporated as part of #1170.
